fastjet/internal/predicates/incircle.go

```diff
@@ -5,110 +5,196 @@
 // |1 x1 y1 x1^2+y1^2|
 // |1 x2 y2 x2^2+y2^2|
 // |1 x3 y3 x3^2+y3^2|
 // |1 x  y  x^2 +y^2 |
-func robustIncircle(x1, y1, x2, y2, x3, y3, x, y *big.Rat) bool {
-	m := [][]*big.Rat{
-		{one, x1, y1, bigAdd(bigMul(x1, x1), bigMul(y1, y1))},
-		{one, x2, y2, bigAdd(bigMul(x2, x2), bigMul(y2, y2))},
-		{one, x3, y3, bigAdd(bigMul(x3, x3), bigMul(y3, y3))},
-		{one, x, y, bigAdd(bigMul(x, x), bigMul(y, y))},
+func robustIncircle(x1, y1, x2, y2, x3, y3, x, y *big.Rat) RelativePosition {
+	m := []*big.Rat{
+		one, x1, y1, bigAdd(bigMul(x1, x1), bigMul(y1, y1)),
+		one, x2, y2, bigAdd(bigMul(x2, x2), bigMul(y2, y2)),
+		one, x3, y3, bigAdd(bigMul(x3, x3), bigMul(y3, y3)),
+		one, x, y, bigAdd(bigMul(x, x), bigMul(y, y)),
 	}
 	det := bigAdd(
 		bigAdd(
 			bigAdd(
-				bigAdd(row(3, 2, 1, 0, false, m), row(3, 1, 2, 0, true, m)),
-				row(2, 1, 3, 0, false, m),
+				bigAdd(rowBig(3, 2, 1, 0, false, m), rowBig(3, 1, 2, 0, true, m)),
+				rowBig(2, 1, 3, 0, false, m),
 			),
 			bigAdd(
-				bigAdd(row(3, 2, 0, 1, true, m), row(3, 0, 2, 1, false, m)),
-				row(2, 0, 3, 1, true, m),
+				bigAdd(rowBig(3, 2, 0, 1, true, m), rowBig(3, 0, 2, 1, false, m)),
+				rowBig(2, 0, 3, 1, true, m),
 			),
 		),
 		bigAdd(
 			bigAdd(
-				bigAdd(row(3, 1, 0, 2, false, m), row(3, 0, 1, 2, true, m)),
-				row(1, 0, 3, 2, false, m),
+				bigAdd(rowBig(3, 1, 0, 2, false, m), rowBig(3, 0, 1, 2, true, m)),
+				rowBig(1, 0, 3, 2, false, m),
 			),
 			bigAdd(
-				bigAdd(row(2, 1, 0, 3, true, m), row(2, 0, 1, 3, false, m)),
-				row(1, 0, 2, 3, true, m),
+				bigAdd(rowBig(2, 1, 0, 3, true, m), rowBig(2, 0, 1, 3, false, m)),
+				rowBig(1, 0, 2, 3, true, m),
 			),
 		),
 	)
-	return det.Cmp(zero) < 0
+	if det.Cmp(zero) < 0 {
+		return Inside
+	}
+	if det.Cmp(zero) == 0 {
+		return On
+	}
+	return Outside
 }
 
-// row is a helper function for robustIncircle
-// m[0][a]*m[1][b]*m[2][c]*m[3][d] - m[0][b]*m[1][a]*m[2][c]*m[3][d] /
+// rowBig is a helper function for robustIncircle
+// If m[row][col] then each row in the determinant calculation is either
+// m[0][a]*m[1][b]*m[2][c]*m[3][d] - m[0][b]*m[1][a]*m[2][c]*m[3][d] or
 // - m[0][a]*m[1][b]*m[2][c]*m[3][d] + m[0][b]*m[1][a]*m[2][c]*m[3][d]
-func row(a, b, c, d int, plus bool, m [][]*big.Rat) *big.Rat {
+func rowBig(a, b, c, d int, plus bool, m []*big.Rat) *big.Rat {
 	if plus {
 		return bigSub(
-			bigMul(bigMul(m[0][b], m[1][a]), bigMul(m[2][c], m[3][d])),
-			bigMul(bigMul(m[0][a], m[1][b]), bigMul(m[2][c], m[3][d])),
+			bigMul(bigMul(m[b], m[a+4]), bigMul(m[c+8], m[d+12])),
+			bigMul(bigMul(m[a], m[b+4]), bigMul(m[c+8], m[d+12])),
 		)
 	}
 	return bigSub(
-		bigMul(bigMul(m[0][a], m[1][b]), bigMul(m[2][c], m[3][d])),
-		bigMul(bigMul(m[0][b], m[1][a]), bigMul(m[2][c], m[3][d])),
+		bigMul(bigMul(m[a], m[b+4]), bigMul(m[c+8], m[d+12])),
+		bigMul(bigMul(m[b], m[a+4]), bigMul(m[c+8], m[d+12])),
 	)
 }
+
+// matIncircle computes the determinant of the matrix using the mat package.
```

Comment:
`matIncircle` only seems to be used for the tests/benchmarks.
put it in that file instead of there.